ass4.cpp

```diff
@@ -0,0 +1,73 @@
+#include <iostream>
+#include <vector>
+#include <deque>
+
+using namespace std;
+
+// First variant of algorithm using recursion
+
+void flood_fill_recursive(vector<vector<bool> > &table, int y, int x) {
+    if (y < 0 || y >= table.size() || x < 0 || x >= table[0].size()) {
+        return;
+    }
+    
+    if (!table[y][x]) {
+        return;
+    }
+    table[y][x] = false;
+    
+    flood_fill_recursive(table, y - 1, x);
+    flood_fill_recursive(table, y + 1, x);
+    flood_fill_recursive(table, y, x - 1);
+    flood_fill_recursive(table, y, x + 1);
+}
+
+// Second variant of algorithm using data structure deque for storing islands
```

Comment:
It was just for practising both recursive and non-recursive algorithms :) I left only recursive for now. Also added the second algorithm Union-Find from Hint 2.